couchbase/src/main/java/org/hibernate/ogm/datastore/couchbase/dialect/backend/json/impl/EntityDocument.java

```diff
@@ -0,0 +1,234 @@
+/*
+ * Hibernate OGM, Domain model persistence for NoSQL datastores
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.ogm.datastore.couchbase.dialect.backend.json.impl;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.regex.Pattern;
+
+import org.hibernate.ogm.datastore.couchbase.logging.impl.Log;
+import org.hibernate.ogm.datastore.couchbase.logging.impl.LoggerFactory;
+import org.hibernate.ogm.datastore.couchbase.util.impl.Identifier;
+import org.hibernate.ogm.model.key.spi.EntityKey;
+import org.hibernate.ogm.model.spi.Tuple;
+
+import com.fasterxml.jackson.annotation.JsonAnyGetter;
+import com.fasterxml.jackson.annotation.JsonAnySetter;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectReader;
+import com.fasterxml.jackson.databind.ObjectWriter;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize.Inclusion;
+
+/**
+ * Contains the information related to a {@link Tuple}. The use of this class is to serialize and deserialize the JSON
+ * stored in CouchBase; Documents have the following structure:
+ *
+ * <pre>
+ * {@code
+ * {
+ *     "_id": "a4jdefe8",
+ *     "_rev": "123",
+ *     "$type": "entity",
+ *     "$table": "Foo",
+ *
+ *     "name": "Bob",
+ *     "login": "dude",
+ *     "homeAddress": {
+ *         "street2": null,
+ *         "street1": "1 avenue des Champs Elysees",
+ *         "country": "France",
+ *         "city": "Paris"
+ *     },
+ * }
+ * }
+ * </pre>
+ *
+ * Implementation note: The entity's properties are stored in a map, with embedded properties being represented by
+ * dot-separated property paths. When (de-)serializing this document from/to JSON via Jackson, this flat representation
+ * is converted into a hierarchical representation using nested maps (if embedded properties are present).
+ *
+ * @author Stawicka Ewa
+ */
+@JsonSerialize(include = Inclusion.NON_NULL)
+@JsonTypeName(EntityDocument.TYPE_NAME)
+public class EntityDocument {
```

Comment:
This looks very similar to the implementation in the _couchdb_ module. Maybe we could extract a common implementation and use it for both?